car-bridge/src/main.rs

```diff
@@ -29,58 +44,203 @@ async fn main() -> Result<(), Box<dyn Error>> {
-            .finalize(),
-    )?;
+    let mut client = MqttMessaging::connect(broker_url, vin.to_owned()).await?;
+    let mut messages = client.subscribe(format!("c2d/{vin}")).await?;
 
-    let mut message_stream = client.get_stream(MQTT_CLIENT_BUFFER_SIZE);
+    let cancellation_token = ctrl_c_cancellation();
 
-    client.connect(ConnectOptionsBuilder::new().mqtt_version(MQTT_VERSION_5).finalize()).await?;
+    let (response_sender, mut response_receiver) = mpsc::channel(PUBLISH_BUFFER);
 
-    let c2d_topic = format!("c2d/{vin}");
-    info!("Subscribing to topic '{c2d_topic}'.");
-    client.subscribe(c2d_topic.clone(), QOS_2).await?;
+    let publish_handle = {
+        // Detach sending the responses from handling the messages to avoid
+        // backpressure and disconnect the client gracefully.
 
-    // Publish a phony message.
-    client.publish(Message::new(c2d_topic, "Example payload", QOS_2)).await?;
+        let cancellation_token = cancellation_token.child_token();
+        spawn(async move {
+            let client = Arc::new(client);
 
-    let cancellation_token = ctrl_c_cancellation();
+            loop {
+                select! {
+                    message = response_receiver.recv() => {
+                        let Some((topic, message)) = message else {
+                            warn!("Response receiver stopped, no more messages will be published.");
+                            break;
+                        };
+
+                        let client = Arc::clone(&client);
+
+                        spawn(async move {
+                            if let Err(e) = client.publish(topic, message).await {
+                                debug!("Error when publishing message: '{:?}'.", e);
+                            }
+                        });
+                    }
+                    _ = cancellation_token.cancelled() => {
+                        debug!("Shutting down the publisher loop.");
+                        break;
+                    }
+                }
+            }
+        })
+    };
 
     loop {
         select! {
-            message = message_stream.next() => {
+            message = messages.next() => {
                 let Some(message) = message else {
                     break;
                 };
-                if let Some(message) = message {
-                    info!("(R) {message:?}");
-                }
-                else {
-                    info!("Connection temporarily lost.");
 
-                    while let Err(err) = client.reconnect().await {
-                        info!("Trying to reconnect: {}.", err);
-                        sleep(Duration::from_secs(5)).await;
-                    }
-                }
+                let mut chariott = chariott.clone();
+                let response_sender = response_sender.clone();
+
+                spawn(async move {
+                    handle_message(&mut chariott, response_sender, message).await;
```

Comment:
The setup here is hinting that `handle_message` ought to consume the cloned `choriott` rather than take a shared reference, since this is what's happening via the async block (or future) anyhow. It should then make the call here look as simple as:

```rust
spawn(handle_message(chariott, response_sender, message));
```

Reply:
Done in 1ff909afe719ecc237a708aab7a20af7a2d9841d